effekt/shared/src/main/scala/effekt/machine/Transformer.scala

```diff
@@ -508,22 +530,60 @@ object Transformer {
    */
 
   class BlocksParamsContext() {
-    var freeParams: Map[Symbol, Environment] = Map()
-    var blockParams: Map[Symbol, Environment] = Map()
+    var info: Map[Symbol, BlockInfo] = Map()
+
+    def definition(id: Symbol): BlockInfo.Definition = info(id) match {
+      case d : BlockInfo.Definition => d
+      case BlockInfo.Parameter(tpe) => sys error s"Expected a function definition, but got a block parameter: ${id}"
+      case BlockInfo.Resumption => sys error s"Expected a function definition, but got a continuation: ${id}"
+    }
+    def blockParams(id: Symbol): Environment = definition(id).blockParams
+    def freeParams(id: Symbol): Environment = definition(id).freeParams
+  }
+  enum BlockInfo {
+    case Definition(freeParams: Environment, blockParams: Environment)
+    case Parameter(tpe: lifted.BlockType)
+    case Resumption
```

Comment:
The idea of this type is that it reflects things we know about block variables in the program.

- `Definition` is a known function (also object? @phischu ?)
- `Parameter` is an unknown function or object
- `Resumption` is a continuation

The calling convention differs between these three, which is why we need to know.